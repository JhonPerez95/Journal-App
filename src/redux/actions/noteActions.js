import Swal from 'sweetalert2';

import { db } from '../../firebase/firebaseConfig';
import { types } from '../types/types';
import { loadNotes } from '../../helpers/loadNotes';

export const activeNote = (id, note) => {
  return {
    type: types.notesActive,
    payload: {
      id,
      ...note,
    },
  };
};

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime(),
    };

    const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);

    dispatch(activeNote(docRef.id, newNote));
  };
};

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes,
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const startSaveNotes = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    if (!note.url) {
      note.url = null;
    }
    const noteUpdate = { ...note };
    delete noteUpdate.id;

    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteUpdate);
    dispatch(refreshNote(note.id, note));
    Swal.fire('Saved', note.title, 'success');
  };
};

export const refreshNote = (id, note) => {
  return {
    type: types.notesUpdate,
    payload: {
      id,
      ...note,
    },
  };
};
