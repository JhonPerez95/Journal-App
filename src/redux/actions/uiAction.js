import { types } from '../types/types';

export const actSetError = (payload) => {
  return {
    type: types.uiSetError,
    payload,
  };
};

export const actRemoveError = () => {
  return {
    type: types.uiRemoveError,
  };
};
