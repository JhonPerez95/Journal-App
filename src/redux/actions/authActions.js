import { types } from '../types/types';
import { firebase, auhtGoogleProvider } from '../../firebase/firebaseConfig';
import { actStartLoading, actFinishLoading } from './uiAction';

import Swal from 'sweetalert2';
import { notesLoguot } from './noteActions';

// Login
export const login = (uid, displayName) => {
  return {
    type: types.login,
    payload: {
      uid,
      displayName,
    },
  };
};

// Logout
export const logout = () => {
  return {
    type: types.logout,
  };
};

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();
    dispatch(logout());
    dispatch(notesLoguot());
  };
};

// Login with Email
export const startEmailLogin = (email, password) => {
  return (dispatch) => {
    dispatch(actStartLoading());
    return firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(actFinishLoading());
      })
      .catch((err) => {
        dispatch(actFinishLoading());
        Swal.fire('Error', err.message, 'error');
      });
  };
};

// Login with Google
export const startGoogleLogin = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(auhtGoogleProvider)
      .then(({ user }) => dispatch(login(user.uid, user.displayName)))
      .catch((err) => Swal.fire('Error', err.message, 'error'));
  };
};

// Register with FireBase
export const startEmailRegister = (email, password, name) => {
  return (dispatch) => {
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
      })
      .catch((err) => Swal.fire('Error', err.message, 'error'));
  };
};
