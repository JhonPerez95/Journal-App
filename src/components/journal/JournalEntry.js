import React from 'react';

import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activeNote } from '../../redux/actions/noteActions';
import PropTypes from 'prop-types';

const JournalEntry = ({ title, body, url, date, id }) => {
  const noteDate = moment(date);
  const dispatch = useDispatch();

  const handleClickEntries = () => {
    dispatch(activeNote(id, { title, body, url, date }));
  };
  return (
    <div className="journal__entry pointer" onClick={handleClickEntries}>
      {url && (
        <div
          className="journal__entry-picture"
          style={{
            backgroundImage: `url(${url})`,
            backgroundSize: 'cover',
          }}
        ></div>
      )}
      <div className="journal__entry-body">
        <p className="journal__entry-title">{title}</p>
        <p className="journal__entry-content">{body}</p>
      </div>
      <div className="journal__entry-date-box">
        <samp>{noteDate.format('dddd')}</samp>
        <h4>{noteDate.format('Do')}</h4>
      </div>
    </div>
  );
};
JournalEntry.propTypes = {
  title: PropTypes.string.isRequired,
  body: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  date: PropTypes.number.isRequired,
  id: PropTypes.string.isRequired,
};

export default JournalEntry;
