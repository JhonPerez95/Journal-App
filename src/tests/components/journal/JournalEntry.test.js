import '@testing-library/jest-dom';
import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store'; //ES6 modules
import JournalEntry from '../../../components/journal/JournalEntry';
import { activeNote } from '../../../redux/actions/noteActions';

// Config Store Redux
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
let store = mockStore(initState);
store.dispatch = jest.fn();

const note = {
  id: 'idaaf65sda5sas5s',
  title: 'Title Testing',
  body: 'Body Testing',
  url: 'https://site.com/testImg.jpg',
  date: 0,
};
const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <JournalEntry {...note} />
    </MemoryRouter>
  </Provider>
);

describe('Test the component <JournalEntry/>', () => {
  test('should show correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should active note', () => {
    wrapper.find('.journal__entry').prop('onClick')();

    expect(store.dispatch).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenCalledWith(
      activeNote(note.id, { ...note })
    );
  });
});
