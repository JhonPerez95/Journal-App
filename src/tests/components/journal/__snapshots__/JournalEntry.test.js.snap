// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the component <JournalEntry/> should show correctly 1`] = `
<div
  className="journal__entry pointer"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://site.com/testImg.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Title Testing
    </p>
    <p
      className="journal__entry-content"
    >
      Body Testing
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <samp>
      Wednesday
    </samp>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
